### ISSUE TITLE
EL9/CentOS9 Basic Support

### ISSUE THREAD

**Issue by Vonng** [Vonng/pigsty#111]
Add support for CentOS 9

**Comment by alemacci** [comment]
I can help on this.
Since CentOS goes Stream (upstream/development branch of RHEL), I think it is possible to evaluate the use of Rocky Linux as CentOS replacement.

**Comment by Vonng** [comment]
It's a great idea to use a template repo file instead of repo_upstreams. Saves lots of config space.

HAProxy rpm used by pigsty are built with https://github.com/Vonng/haproxy-rpm.

Which enables built-in Prometheus metrics support. Maybe I can compile an El8/EL9 version and upgrade to v2.6

**Comment by alemacci** [comment]
Another step forward.

I am testing OL8 also, which is working well.

Another issue is when we use '**yum install --downloadonly**' cmd for downloading repo packages on **_./roles/repo/tasks/main.yml_**, il will not download packages that have already been installed on control node.

I think we can rely on repotrack command, which will do both jobs and works well on newest EL builds (on EL8/9 it became an alias of 'dnf download' command).

While specifing '--arch x86_64,noarch' switch, i686 packages can be excluded. This behavior doesn't work on EL7 due to https://bugzilla.redhat.com/show_bug.cgi?id=1150115 (but code can be patched https://getridbug.com/unix-linux/download-all-dependencies-with-yumdownloader-even-if-already-installed/)

```
    # download packages from upstream yum repo
    - name: Download repo packages and their dependencies
      tags: repo_pkg
      environment: "{{ proxy_env }}"
      command:
        cmd: repotrack --arch x86_64,noarch {{ item }}
        chdir: "{{ nginx_home }}/{{ repo_name }}"
      with_items: "{{ repo_packages }}"
```

**Comment by Vonng** [comment]
Maybe I can remove i686 packages when making pkg.tgz when using el7 ...

**Comment by Vonng** [comment]
What about using a series of a predefined files instead of if-else templates.

Just like `pg_conf`?

And we may get a list of combinations such as :

* centos7
* centos8
* centos9

* centos7-china (mirror)
* centos8-china (mirror)
* centos9-china (mirror)

* rocky7
* rocky8
* rocky9

* oracle7
* oracle8
* oracle9

* alma8
* alma9

or we can use both?

* centos.repo
* oracle.repo
* rocky.repo
* alma.repo

and 

```yaml
if os_major_version == 7:...
if os_major_version == 8:...
if os_major_version == 9:...
```

**Comment by alemacci** [comment]
I think the second solution is more sleek.
Maybe we can add a var like `os_region` or something to be like 'europe' or 'china' and configure repo mirrors based on that?

**Comment by Vonng** [comment]
brilliant idea. so we got one template file for each OS Name.
and distinguish different repos by `os_major_build` & `os_region`.

What if we make os_name & os_major_build as config parameters.
And optionally detect and set them during runtime?

* os_name: CentOS
* os_version: 7
* repo_region: default|china|europe|...

**Comment by alemacci** [comment]
Ok!

To sum up:

- Add os_name, os_version, repo_region variables
- Move 'Gather OS specs' runtime from environ to repo but use them optionally (i.e. when os_name & os_version are not defined or os_name is set to 'auto')
- One template file for each OS based on os_version & repo_region

Also, I think we can use os_version var to make dynamic repo_url_packages, for example:
      - https://github.com/Vonng/haproxy-rpm/releases/download/v2.6.1/haproxy-2.6.1-1.el${os_version}.x86_64.rpm

**Comment by alemacci** [comment]
Updates:

- ntpd support is removed from EL8 onwards. Shall we consider removing it too in favour of Chrony?
- dynamic repo_url_packages urls are working well;
- some packages such as etcd and pgloader don't exists on official EL/8 repos;
- I cleaned up some python packages from repo_packages as they will be downloaded anyway as dependencies; for that sake, I will upload a new `pigsty-el.yml` conf example file;

There are some slight caveats in Ansible version on various OSes. For example, Oracle 8 EPEL will still install Ansible v2.9, whereas Rocky/Alma/CentOS 8 EPEL will install Ansible v2.12 that needs python 3.8 to work properly (control node only).

I will update ASAP

**Comment by Vonng** [comment]
Everything seems fine and works on EL8/EL9 Rocky/CentOS/Alma/Oracle. It will be released on v1.6.0. 

EL8 example config: https://github.com/Vonng/pigsty/blob/master/files/conf/pigsty-el8.yml
EL9 excample config: https://github.com/Vonng/pigsty/blob/master/files/conf/pigsty-el9.yml

And, thanks for your great contribution 🎉 @alemacci
